File: java/smr-collections-common/src/test/java/net/knego/hiperf/smrcollections/GoldenTest.java
package net.knego.hiperf.smrcollections;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import net.knego.hiperf.common.SmrConfig;
import org.junit.jupiter.api.Test;

class GoldenTest {
    @Test
    void crossLanguageGoldenBytes() throws Exception {
        // Gradle test working dir is the subproject dir (java/smr-collections-common).
        Path golden = Path.of("..", "..", "rust", "smr-collections", "testdata", "golden_snapshot.bin");
        byte[] want = Files.readAllBytes(golden);
        SmrConfig c = new SmrConfig(4096, 64, 1, 0, 2000, 0, 0);
        Book b = new Book(c);
        Workload.SplitMix rng = new Workload.SplitMix(Workload.SEED);
        Workload.Insert ins = new Workload.Insert();
        for (int i = 0; i < c.steady(); i++) {
            Workload.nextInsert(rng, i, c.levels(), c.tick(), c.priceMin(), ins);
            b.insert(ins.orderId, ins.price, ins.qty, ins.side);
        }
        Snapshotter s = new Snapshotter(4 * 1024 * 1024);
        int len = s.encode(b);
        byte[] got = Arrays.copyOf(s.backing(), len);
        assertArrayEquals(want, got, "java snapshot bytes must equal the rust golden");
    }
}

File: go/internal/smrcoll/golden_test.go
package smrcoll

import (
	"bytes"
	"os"
	"testing"

	"github.com/peterknego/hi-perf-cmp/go/internal/bench"
)

func TestCrossLanguageGoldenBytes(t *testing.T) {
	golden, err := os.ReadFile("../../../rust/smr-collections/testdata/golden_snapshot.bin")
	if err != nil {
		t.Fatalf("read golden (run the Rust R4 export first): %v", err)
	}
	c := bench.SmrConfig{Cap: 4096, Levels: 64, Tick: 1, PriceMin: 0, Steady: 2000, Warmup: 0, Iters: 0}
	s := NewSnapshotter()
	got := s.Encode(buildBook(c, c.Steady))
	if !bytes.Equal(got, golden) {
		t.Fatalf("go snapshot bytes differ from rust golden (len go=%d rust=%d)", len(got), len(golden))
	}
}
